keps/sig-node/3857-rro-mounts/README.md

```diff
@@ -0,0 +1,1085 @@
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+Constraints: needs runc >= 1.1 && kernel >= 5.12.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+Risk: a manifest with `recursiveReadOnly: Enabled` will not work on old distros (See "Constraints" above).
+
+Mitigation: set `recursiveReadOnly: IfPossible`.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+
+### Core API
+Add `RecursiveReadOnly: (Disabled|IfPossible|Enabled)` to the [`VolumeMount`](https://github.com/kubernetes/kubernetes/blob/v1.26.1/pkg/apis/core/types.go#L1854-L1880) struct.
+
+A pod manifest will look like this:
+```yaml
+spec:
+  volumes:
+    - name: foo
+      hostPath:
+        path: /mnt
+        type: Directory
+  containers:
+  - volumeMounts:
+    - mountPath: /mnt
+      name: foo
+      mountPropagation: None
+      readOnly: true
+      # NEW
+      recursiveReadOnly: IfPossible
+```
+
+See the comment lines in the diff below for the constraints of the `VolumeMount` options:
+```diff
+diff --git a/pkg/apis/core/types.go b/pkg/apis/core/types.go
+index 04ffd250b61..ae847f2f4f8 100644
+--- a/pkg/apis/core/types.go
++++ b/pkg/apis/core/types.go
+@@ -1858,6 +1858,24 @@ type VolumeMount struct {
+ 	// Optional: Defaults to false (read-write).
+ 	// +optional
+ 	ReadOnly bool
++	// RecursiveReadOnly specifies recursive-readonly mode.
++	//
++	// 1. If ReadOnly is false, RecursiveReadOnly must be unspecified.
++	// 2. If ReadOnly is true:
++	//   2.1. If RecursiveReadOnly is unspecified:
++	//        2.1.1. if it belongs to a Pod being created, it is initialized to Disabled.
++	//        2.1.2  if it belongs to a PodSpec under Deployment, Job, etc., it remains unspecified
++	//               (and will be set to Disabled eventually, when the Pod is created).
++	//   2.2. If RecursiveReadOnly is set to Disabled, the mount is not made recursively read-only.
++	//   2.3. If RecursiveReadOnly is set to IfPossible, the mount is made recursively read-only,
++	//        if it is supported by the runtime.
++	//        MountPropagation must be None or unspecified (which defaults to None).
++	//   2.4. If RecursiveReadOnly is set to Enabled, the mount is made recursively read-only.
```

Comment:
Updated
```
+       // RecursiveReadOnly specifies recursive-readonly mode.
+       //
+       // 1. If ReadOnly is false, RecursiveReadOnly must be unspecified.
+       // 2. If ReadOnly is true:
+       //   2.1. If RecursiveReadOnly is unspecified:
+       //        2.1.1. if it belongs to a Pod being created, it is initialized to Disabled.
+       //        2.1.2  if it belongs to a PodSpec under Deployment, Job, etc., it remains unspecified
+       //               (and will be set to Disabled eventually, when the Pod is created).
+       //   2.2. If RecursiveReadOnly is set to Disabled, the mount is not made recursively read-only.
+       //   2.3. If RecursiveReadOnly is set to IfPossible, the mount is made recursively read-only,
+       //        if it is supported by the runtime.
+       //        If it is not supported by the runtime, the mount is not made recursively read-only.
+       //        MountPropagation must be None or unspecified (which defaults to None).
+       //   2.4. If RecursiveReadOnly is set to Enabled, the mount is made recursively read-only.
+       //        If it is not supported by the runtime, this will result in an error.
+       //        MountPropagation must be None or unspecified (which defaults to None).
+       //   2.5. If RecursiveReadOnly is set to unknown value, it will result in an error.
+       //
+       // +featureGate=RecursiveReadOnlyMounts
+       // +optional
+       RecursiveReadOnly *RecursiveReadOnlyMode
```